### ISSUE TITLE
Connect to MongoDB error: Connection to ... successful, but could not connect to DB.

### ISSUE THREAD

**Issue by anki-code** [metabase/metabase#2722]
Hi!
I'm trying to add remote MongoDB 3.2.6 (**auth by login and password**) to Metabase 0.17.1.
Metabase is running from the JAR on Ubuntu 14.04. Java version:

```
$ java -version
java version "1.7.0_101"
OpenJDK Runtime Environment (IcedTea 2.6.6) (7u101-2.6.6-0ubuntu0.14.04.1)
OpenJDK 64-Bit Server VM (build 24.95-b01, mixed mode)
```

Connect from `mongo` client **by login and password** works successfully but when I try to add MongoDB in Metabase Admin Panel I get the error `Connection to ... successful, but could not connect to DB.` and trace below:

```
06-09 12:09:52 DEBUG mongo.util :: << OPENED NEW MONGODB CONNECTION >>
06-09 12:09:55 ERROR metabase.driver :: Failed to connect to database: com.mongodb.MongoTimeoutException: Timed out after 3000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=UNKNOWN, servers=[{address=...here my ip...:26017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketWriteException: Exception sending message}, caused by {javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake}, caused by {java.io.EOFException: SSL peer shut down incorrectly}}]
06-09 12:09:55 DEBUG mongo.util :: << OPENED NEW MONGODB CONNECTION >>
```

How to fix the problem?

**Comment by anki-code** [comment]
I've got the same error on java 1.8
```
openjdk version "1.8.0_91"
OpenJDK Runtime Environment (build 1.8.0_91-8u91-b14-0ubuntu4~14.04-b14)
OpenJDK 64-Bit Server VM (build 25.91-b14, mixed mode)
```

**Comment by anki-code** [comment]
I've built JAR from `master` (ver 0.18) sources and get the same error:
```
06-09 20:03:12 ERROR metabase.driver :: Failed to connect to database: com.mongodb.MongoTimeoutException: Timed out after 3000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=UNKNOWN, servers=[{address=...:26017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketWriteException: Exception sending message}, caused by {javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake}, caused by {java.io.EOFException: SSL peer shut down incorrectly}}]

06-09 20:03:12 DEBUG metabase.middleware :: POST /api/setup/validate 400 (3 s)
{:message
 "com.mongodb.MongoTimeoutException: Timed out after 3000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=UNKNOWN, servers=[{address=...:26017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketWriteException: Exception sending message}, caused by {javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake}, caused by {java.io.EOFException: SSL peer shut down incorrectly}}]"}
```

**Comment by anki-code** [comment]
Any thoughts?

**Comment by camsaul** [comment]
Hi @anki-code, sorry for the long delay before replying. Not sure why you're seeing that issue. You're able to connect with the same details using the `mongo` command line client, right?

Can you try connecting to any other Mongo servers if possible?

**Comment by camsaul** [comment]
@anki-code, can you connect to the database from the mongo command line client?

**Comment by camsaul** [comment]
@anki-code?

**Comment by anki-code** [comment]
I've tried to use metabase-0.20 but still have the error:

`Failed to connect to database: com.mongodb.MongoTimeoutException: Timed out after 3000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=UNKNOWN, servers=[{address=mymongoserver:28017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketWriteException: Exception sending message}, caused by {javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake}, caused by {java.io.EOFException: SSL peer shut down incorrectly}}]`

@camsaul 
> You're able to connect with the same details using the mongo command line client, right?

Yes.

> Can you try connecting to any other Mongo servers if possible?

I've tried local mongo server on the same host with Metabase. It works.

**Comment by anki-code** [comment]
@camsaul, here is mongo config:

```
security:
 authorization: enabled
 keyFile: /mnt/data/mdb.keyfile
 clusterAuthMode: keyFile

net:
 bindIp: 0.0.0.0
 maxIncomingConnections: 1000000
 port: 28017
 http:
   enabled: false

storage:
 directoryPerDB: true
 dbPath: /mnt/data/db
 journal:
   enabled: true
 engine: wiredTiger

systemLog:
 destination: file
 logAppend: true
 path: /var/log/mongodb/mongod.log
 logRotate: reopen
processManagement:
 fork:  false

replication:
 replSetName: p-prod
 secondaryIndexPrefetch: all
 oplogSizeMB: 2048
```

**Comment by camsaul** [comment]
@anki-code Which parts of that config did you change from the default settings? That should help us pinpoint what's going on

**Comment by anki-code** [comment]
@camsaul nothing special. If you will send me special jar file with SSL debug messages I could be test it ASAP.

**Comment by camsaul** [comment]
@anki-code does this sound like the same issue as #3877 ?

**Comment by anki-code** [comment]
It could be

**Comment by camsaul** [comment]
@anki-code are you using a self-signed SSL certificate?

**Comment by anki-code** [comment]
@camsaul no. We don't use SSL. Connect to db from mongo client by login and password works successfully without any certificates.

Could MongoDB driver in Metabase require SSL? May be the SSL handshake error appears when Metabase try to connect via SSL but MongoDB server doesn't use SSL.

**Comment by camsaul** [comment]
@anki-code Ok, then it sounds like your issue is not the same issue as #3877 :joy:

Metabase doesn't require an SSL connection for Mongo but it tries to connect with SSL first before trying without SSL.

**Comment by anki-code** [comment]
@camsaul could you please tell me how to switch SSL off in Metabase? I can do it and tell you about result.

**Comment by camsaul** [comment]
@anki-code you don't need to switch off SSL. Metabase will automatically try your connect both with and without SSL.

**Comment by anki-code** [comment]
@camsaul my thought was about Exception when Metabase tries SSL. If SSL connection going first Metabase could throw the Exception and exit without trying not SSL connection. This case may be resolved if we'll switch SSL off and the connection could be established without SSL. Sorry if it is my fantasy. I'm not see the code.

**Comment by davidmir** [comment]
I have the same issue. Metabase gives an exception when trying to connect to MongoDB and don't tries again without SSL. How can we solve this issue?

**Comment by cgranier** [comment]
I was having the same issue, and this solved it:

I'm running a docker image of metabase and a separate docker image of mongo. Trying to connect to mongo using localhost:27017 from within metabase was giving me the exception posted above.

Thanks to [this](http://stackoverflow.com/questions/33336773/connecting-to-mongo-docker-container-from-host) SO post, I ran:

`docker inspect --format '{{ .NetworkSettings.IPAddress }}' mongo-database`

and got back an IP address different than localhost. I used that IP address on the metabase add database form and it connected!

**Comment by camsaul** [comment]
@anki-code @davidmir did @cgranier's fix work for you?

**Comment by camsaul** [comment]
@anki-code ?

**Comment by anki-code** [comment]
@camsaul I'm not using docker.

**Comment by davidmir** [comment]
I think the problem is not solved with this solution.

**Comment by Sohojoe** [comment]
I have the same problem... Connection to 'x.x.x.x:27001' successful, but could not connect to DB.

our db is hosted in a data centre - I'm able to connect using Studio 3T and MongoDB Compass

**Comment by camsaul** [comment]
I have a feeling we may need to allow [more configuration options for authorization](http://api.mongodb.com/java/3.2/?com/mongodb/MongoClientURI.html). (I think we might need a way to configure `authSource`/`authMechanism`?)

@anki-code @davidmir @Sohojoe can you give me a sense of what your connection strings/options look like when you connect to other tools? Do you have to specify `authSource` or anything like that that isn't an option in Metabase?

**Comment by camsaul** [comment]
@anki-code @davidmir @Sohojoe do you guys set `authMechanism` or `authSource` or anything like that? In #4971 I added a field to specify additional connection options for Mongo so if you guys need to set those you'll be able to do so now

**Comment by camsaul** [comment]
(Possibly fixed by #4971)

**Comment by bluemana** [comment]
I was having the same issue and I solved it following the steps at https://docs.docker.com/docker-for-mac/networking/#use-cases-and-workarounds (I'm running the Metabase docker container on macOS):

1. Attach an unused IP to the `lo0` interface on the MongoDB host. For example:

```
sudo ifconfig lo0 alias 10.200.10.1/24
```

2. Make sure `mongod` is listening to that IP or `0.0.0.0`:

```
net:
  bindIp: 10.200.10.1
```
Metabase should now be able to connect to MongoDB.

**Comment by camsaul** [comment]
@anki-code @davidmir @Sohojoe do you guys set authMechanism or authSource or anything like that? In #4971 I added a field to specify additional connection options for Mongo so if you guys need to set those you'll be able to do so now

If not does @bluemana's fix work for you?

**Comment by anki-code** [comment]
@camsaul I've built the latest master and tried to use various of "Additional Mongo connection string options". No result.

When I try to connect to DB I see two tries of connection in the log (I suppose the first is SSL and the second is not SSL):

```
07-19 12:40:46 DEBUG mongo.util :: << OPENED NEW MONGODB CONNECTION >>
07-19 12:40:49 DEBUG mongo.util :: << CLOSED MONGODB CONNECTION >>
07-19 12:40:49 ERROR metabase.driver :: Failed to connect to database: com.mongodb.MongoTimeoutException: Timed out after 3000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=UNKNOWN, servers=[{address=mdbhst0.prod.hosts:30021, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketWriteException: Exception sending message}, caused by {javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake}, caused by {java.io.EOFException: SSL peer shut down incorrectly}}]
07-19 12:40:49 DEBUG metabase.middleware :: POST /api/setup/validate 400 (3 s) (0 DB calls)
{:message
 "com.mongodb.MongoTimeoutException: Timed out after 3000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=UNKNOWN, servers=[{address=mdbhst0.prod.hosts:30021, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketWriteException: Exception sending message}, caused by {javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake}, caused by {java.io.EOFException: SSL peer shut down incorrectly}}]"}


07-19 12:40:49 DEBUG mongo.util :: << OPENED NEW MONGODB CONNECTION >>
07-19 12:40:50 DEBUG mongo.util :: << CLOSED MONGODB CONNECTION >>
07-19 12:40:50 DEBUG metabase.middleware :: POST /api/setup/validate 400 (128 ms) (0 DB calls)
{:errors {:dbname "Connection to 'mdbhst0.prod.hosts:30021' successful, but could not connect to DB."}}
```

We can ignore the first connection error (SSL) in this issue. But the reason of the second error ("could not connect to DB") is the great question.

**Comment by camsaul** [comment]
@anki-code strange. I don't know what it's not working 😞

**Comment by anki-code** [comment]
@camsaul we debug the settings and the results are:

1. The error `nnection to 'mdbhst0.prod.hosts:30021' successful, but could not connect to DB.` appears when user haven't `listCollections` and `dbStats` permissions. **We've added this permissions and resolve the error**. (It's not so good for our security because in our mongodb we don't want to show some collections to the users but with permission `listCollections` they can see the names.)

2. The error 
```
Failed to connect to database: com.mongodb.MongoTimeoutException: Timed out after 3000 ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}. Client view of cluster state is {type=UNKNOWN, servers=[{address=mdbhst0.prod.hosts:30021, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketWriteException: Exception sending message}, caused by {javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake}, caused by {java.io.EOFException: SSL peer shut down incorrectly}
```
appears on first attempt when Metabase try to make SSL connection to mongodb server without enabling SSL. This error appears even if we add `ssl=false` in connection settings. But this error isn't influence to next connection attempt without SSL.

As result my problem resolved :)

**Comment by richardlt** [comment]
Hi,

Looks like I have the same issue (com.mongodb.MongoTimeoutException: Timed out after 3000 ms) when trying to connect to a Mongo database with a ssl tunnel. 
The auth is disabled on this database and ssl is not setup. I tried a lot of thing running Metabase inside Docker or with the mac App. It works with the same setup in Robo 3T. I really don't understand what's wrong. Do you have an idea ?

Thanks.

**Comment by tpakeman** [comment]
phhhhhrp

**Comment by mralexpopa** [comment]
Any updates on the issue? I'm having the same problem

**Comment by manojap** [comment]
I have same 3000 timeout issue , I am digging

**Comment by kalyandas** [comment]
I am facing the same issue

**Comment by ujwaljaglan9211** [comment]
i'm in trouble in proper working of mongodb on windows 10,
i followed all steps to use mongodb version 3.6,
i run mongod.exe command ang get,waiting for connection on port 27017...
after i run mongo.exe in other cmd then nothing happens in this cmd...
i also used studio3t and robomongo...but i found negative responses.
/*************************************
2018-10-29T12:31:20.524+0530 I NETWORK  [initandlisten] waiting for connections on port 27017
2018-10-29T12:31:34.545+0530 I NETWORK  [listener] connection accepted from 127.0.0.1:53287 #1 (1 connection now open)
2018-10-29T12:31:51.548+0530 I NETWORK  [conn1] end connection 127.0.0.1:53287 (0 connections now open)
2018-10-29T12:31:55.054+0530 I NETWORK  [listener] connection accepted from 127.0.0.1:53290 #2 (1 connection now open)
2018-10-29T12:32:12.055+0530 I NETWORK  [conn2] end connection 127.0.0.1:53290 (0 connections now open)
*****************************************************************************/
This happens when i connect via studio3t or robomongo
Please help me,,Thanks in advance
mail:- ujwal@sparkeighteen.com

**Comment by tjramage** [comment]
@anki-code @camsaul — I had this exact issue. I have a MongoDB deployment with compose.io. I've spent the whole day trying to figure it out... I downloaded the Mac desktop client to initially try out Metabase and see if I like it. I plugged in my production connection string with a newly created (read-only) user and, hey presto! It worked absolutely fine.

I then decided I would setup a proper server, running within a Docker container. However, it fails — using the exact same connection details. It can't connect to the database, complaining about SSL exceptions.

I finally stumbled across @BoLaMN's comment in this issue: https://github.com/metabase/metabase/issues/6678#issuecomment-388415723

IT WORKED! I'm now using his custom `bolamn/metabase-head:0.30-snap-debug` image and I'm finally running. Obviously I would much prefer to be using the 0.31 release, though.

Can we make the necessary changes to support these kinds of setups with Compose.io and Mongo Atlas, that use shared clusters??
